python/mxnet/gluon/trainer.py

```diff
@@ -158,3 +158,37 @@ def step(self, batch_size, ignore_stale_grad=False):
                 if not ignore_stale_grad or arr._fresh_grad:
                     upd(i, grad, arr)
                     arr._fresh_grad = False
+
+    def save_states(self, fname):
+        """Saves trainer states (e.g. optimizer, momentum) to a file.
+
+        Parameters
+        ----------
+        fname : str
+            Path to output states file.
+        """
+        assert self._optimizer is not None
+
+        if self._update_on_kvstore:
+            self._kvstore.save_optimizer_states(fname, dump_optimizer=True)
+        else:
+            with open(fname, 'wb') as fout:
+                fout.write(self._updaters[0].get_states(dump_optimizer=True))
+
+    def load_states(self, fname):
+        """Loads trainer states (e.g. optimizer, momentum) from a file.
+
+        Parameters
+        ----------
+        fname : str
+            Path to input states file.
+        """
+        if self._update_on_kvstore:
+            self._kvstore.load_optimizer_states(fname)
+            self._optimizer = self._kvstore._updater.optimizer
+        else:
+            with open(fname, 'rb') as f:
+                states = f.read()
+            for updater in self._updaters:
+                updater.set_states(states)
+            self._optimizer = self._updaters[0].optimizer
```

Comment:
Do different self._updaters have different optimizer?

Reply:
Should be the exact same instance. The calling sequence are the following:
```python
# Trainer.__init__ calls Trainer._init_optimizer:
    def _init_optimizer(self, optimizer, optimizer_params):
        param_dict = {i: param for i, param in enumerate(self._params)}
        if isinstance(optimizer, opt.Optimizer):
            assert not optimizer_params, \
                "optimizer_params must be None if optimizer is an instance of " \
                "Optimizer instead of str"
            self._optimizer = optimizer
            self._optimizer.param_dict = param_dict
        else:
            self._optimizer = opt.create(optimizer, param_dict=param_dict,
                                         **optimizer_params)

        self._updaters = [opt.get_updater(self._optimizer) \
                            for _ in self._contexts]

# and optimizer.get_updater is:
def get_updater(optimizer):
    return Updater(optimizer)

# and Updater __init__ looks like this:
class Updater(object):
    """Updater for kvstore."""
    def __init__(self, optimizer):
        self.optimizer = optimizer
```

Reply:
But after load_state they would be different

Reply:
Got it. Let me set them to be the same instance in the for loop.